### DISCUSSION TITLE
Twitch provider scope customisation

### DISCUSSION THREAD

**Discussion by khendrikse** [nextauthjs/next-auth#2810] [Help] [answered]
Hi! 

I am trying to set a specific scope for the Twitch provider, but I keep getting a 401 with "Missing scope: user:read:follows" when I use the `access_token` that I receive after logging in. The token does work for all endpoints that don't need that scope.

I thought this is how I was supposed to set scopes for this specific provider, but I am wondering if I might be doing something wrong:

```jsx
import TwitchProvider from `next-auth/providers/twitch`
...
providers: [
  TwitchProvider({
    clientId: process.env.TWITCH_CLIENT_ID,
    clientSecret: process.env.TWITCH_CLIENT_SECRET,
    scope: 'user:read:follows'
  })
]
...
```

**Comment by balazsorban44** [comment, marked as answer]
Using v4 I assume. Have a read of these:

https://next-auth.js.org/getting-started/upgrade-v4#providers
https://github.com/nextauthjs/next-auth/pull/2411#issue-693918157
https://next-auth.js.org/configuration/providers/oauth-provider#options


TLDR;

```js
import TwitchProvider from `next-auth/providers/twitch`
...
providers: [
  TwitchProvider({
    clientId: process.env.TWITCH_CLIENT_ID,
    clientSecret: process.env.TWITCH_CLIENT_SECRET,
    authorization: {
      params: {
        scope: 'user:read:follows'
      }
    }
  })
]
...

```

**Reply by khendrikse** [reply]
I’m afraid I did not see that. Thanks for the quick reply!

**Reply by khendrikse** [reply]
Strange, I changed this, but now I get an `id_token not present in TokenSet` error. Is there a way for me to get access to the data in the `OAUTH_CALLBACK_ERROR` to debug? Or is this something that you recognise. 

I tried making sure I am using the right node version (15) and I am on `"next-auth": "4.0.0-beta.2",`

I think it may have something to do with `openid-client` as this is where the error is popping op from:

```
[next-auth][error][CALLBACK_OAUTH_ERROR] 
https://next-auth.js.org/errors#callback_oauth_error id_token not present in TokenSet TypeError: id_token not present in TokenSet
    at Client.validateIdToken (/Users/user/projects/project/node_modules/openid-client/lib/client.js:685:15)
    at Client.callback (/Users/user/projects/project/node_modules/openid-client/lib/client.js:469:18)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async oAuthCallback (webpack-internal:///./node_modules/next-auth/server/lib/oauth/callback.js:87:16)
    at async Object.callback (webpack-internal:///./node_modules/next-auth/server/routes/callback.js:50:11)
    at async NextAuthHandler (webpack-internal:///./node_modules/next-auth/server/index.js:171:18)
    at async eval (webpack-internal:///./node_modules/next-auth/server/index.js:251:32)
    at async Object.apiResolver (/Users/user/projects/project/node_modules/next/dist/server/api-utils.js:101:9)
    at async DevServer.handleApiRequest (/Users/user/projects/project/node_modules/next/dist/server/next-server.js:760:9) {
  name: 'OAuthCallbackError'
}
```

**Reply by balazsorban44** [reply]
I think you will need to add the `openid` scope.

> Verify that a scope parameter is present and contains the `openid` scope value. (If no `openid` scope value is present, the request may still be a valid OAuth 2.0 request, but is not an OpenID Connect request.)

https://openid.net/specs/openid-connect-core-1_0.html#AuthRequestValidation

The default TwitchProvider config assumes it is used as an OpenID Connect provider

**Reply by khendrikse** [reply]
Again: thanks!

**Reply by khendrikse** [reply]
@balazsorban44 is there a way in which I can opt-out of using OpenId and just get a normal OAuth token? (I need a refresh token which the OIDC flow of Twitch does not support).

This is all quite new to me and even though I really appreciate the amazing docs I can't seem to work it out
